cnd/src/network/mod.rs

```diff
@@ -128,6 +131,68 @@ impl Swarm {
             local_peer_id,
         })
     }
+
+    /// Takes all the data we know about a 'created' swap.  Executes the
+    /// appropriate communication protocols in order to send data to the peer
+    /// and receive data from the peer (remaining swap parameters) in order to
+    /// finalize this swap.
+    ///
+    /// This works for all supported combinations on Han, HALight, and Herc20.
+    pub async fn finalize_created_swap<A, B>(
+        &mut self,
+        _created: CreatedSwapParamaters<A, B>,
+    ) -> anyhow::Result<RemainingSwapParameters<A, B>> {
+        // `comit.behaviour` is the NetworkBehaviour that implements the new
+        // communication protocols.
+
+        // let mut _swarm = self.swarm.lock().await;
+        // swarm.behaviour.finalize_created_swap(_created).await?
+
+        unimplemented!()
+    }
+}
+
+/// Holds all the data that needs to be shared with Bob by Alice to take a swap
+/// from 'created' to 'finalized'.
+///
+/// For Han-Han, Han-Herc20, Herc20-Han
+/// - alpha_identity: refund identity for Alice
+/// - beta_identity: redeem identity for Alice
+///
+/// For Han-HALight
+/// - alpha_identity: refund identity for Alice (i.e., Ethereum address)
+/// - beta_identity: lightning identity for Alice
+///
+/// For HALight-Han
+/// - alpha_identity: lightning identity for Alice
+/// - beta_identity: redeem identity for Alice (i.e., Ethereum address)
+#[derive(Debug)]
+pub struct CreatedSwapParamaters<A, B> {
```

Comment:
I've got some design-challenging questions for you:

- How will Bob be able to call this API? He doesn't have knowledge of the secret hash before the communication protocol is finished. Thinking this further, we are we forcing the caller to generate a secret hash? We can we not do this inside?
- The SwapId is only established as part of the communication protocol, which SwapId is this referring to?
- The identities are tricky. We need to generate a transient one for Bitcoin due to the nature of our HTLC and the complicated redeem/refund process. For Ethereum, we can pass the one through that the user gave us on the API. Can/should we take some of the caller's work (generating a transient Bitcoin identity) away by doing this inside? What does the API look like if we don't explicitly pass identities in?

Reply:
> How will Bob be able to call this API?

If Bob should not call this API, then it the one he should be calling is missing in this PR.

Reply:
> I've got some design-challenging questions for you:

Thought provoking review, thanks.
 
> * How will Bob be able to call this API? He doesn't have knowledge of the secret hash before the communication protocol is finished. 

I totally did not think about the Bob side of things, forgetting that Bob would need to call _back_ into the swarm after receiving the `swap_digest`.  Please see new function and improved stucts.

> Thinking this further, we are we forcing the caller to generate a secret hash? We can we not do this inside?

The secret hash will need to be accessible from outside the swarm, will it not?  So I think it is better to pass it in than pass it out.  

> * The SwapId is only established as part of the communication protocol, which SwapId is this referring to?

Touche

> * The identities are tricky. We need to generate a transient one for Bitcoin due to the nature of our HTLC and the complicated redeem/refund process. For Ethereum, we can pass the one through that the user gave us on the API. Can/should we take some of the caller's work (generating a transient Bitcoin identity) away by doing this inside? What does the API look like if we don't explicitly pass identities in?

As with the SwapId I feel that the implementation details (generating the transient Bitcoin identity) should be done outside of the network module to keep that module cleaner.

Thanks for your effort @thomaseizinger , I did not think this through fully and hence loaded your brain up more than was necessary.

Reply:
> I totally did not think about the Bob side of things, forgetting that Bob would need to call back into the swarm after receiving the swap_digest. Please see new function and improved stucts.

Why does he need to call _back_ into it? Have you considered that the the controller may not (really) care whether they are Alice or Bob and just call the same function on the swarm and let it figure out what to do next, i.e. for Alice sending an announce and for Bob waiting for an announce?
The role information would obviously need to be communicated somehow but that can also be a parameter instead of two different function calls.

> The secret hash will need to be accessible from outside the swarm, will it not? So I think it is better to pass it in than pass it out.

So will all the other parameters that are received from the other side, no?
Why should Alice pass it in and Bob only receive it as an output? In the controller, I would like to avoid this kind of conditional logic and just call the Swarm with the minimally necessary data to start the communication protocol, regardless of whether it is Alice or Bob.

> As with the SwapId I feel that the implementation details (generating the transient Bitcoin identity) should be done outside of the network module to keep that module cleaner.

Let's pass them in for now and see what it evolves too. I feel like in the long-term, it would be better to have this module handle that logic though. It looks cleaner to me to output all the protocol-specific information in the end instead of doing parts upfront and parts in the end.

Reply:
> Thanks for your effort @thomaseizinger , I did not think this through fully and hence loaded your brain up more than was necessary.

No worries man :)